packages/server-wallet/src/onchain-service/utils.ts

```diff
@@ -0,0 +1,46 @@
+import {Evt} from 'evt/lib/types';
+import pino from 'pino';
+
+import {Values} from '../errors/wallet-error';
+
+export abstract class BaseError extends Error {
+  static readonly errors = {
+    OnchainError: 'OnchainError',
+    TransactionError: 'TransactionError',
+    StorageError: 'StorageError',
+  } as const;
+
+  readonly context: any;
+
+  static readonly knownErrors: {[key: string]: string};
+
+  static isKnownErr(errorMessage: string, knownErrors: string[]): boolean {
+    const idx = knownErrors.findIndex(known => errorMessage.includes(known));
+    return idx !== -1;
+  }
+
+  abstract readonly type: Values<typeof BaseError.errors>;
+  static readonly reasons: {[key: string]: string};
+  constructor(reason: Values<typeof BaseError.reasons>, public readonly data: any = undefined) {
+    super(reason);
+    this.context = data;
+  }
+}
+
+// Adds a handler to an evt instance and returns the result
+// based on the input arguments
+export const addEvtHandler = (
+  evt: Evt<any>,
+  callback: (event: any) => void | Promise<void>,
+  filter?: (event: any) => boolean,
+  timeout?: number
+): Evt<any> | Promise<any> => {
+  if (filter && timeout) {
+    return evt.attach(filter, timeout, callback);
+  } else if (filter) {
+    return evt.attach(filter, callback);
+  }
+  return evt.attach(callback);
```

Comment:
This looks a bit nasty, partially due to the signature of `evt.attach`. In particular, it seems to be missing the casae where `timeout` is truthy, and `filter` is not.

I can think of two solutions:
1. overwrite default values with the provided values, if they are truthy (preferable, if that makes sense)
2. the following suggestion
```suggestion
  const attachArgs = [];
  if (filter) attachArgs.push(filter);
  if (timeout) attachArgs.push(timeout);
  attachArgs.push(callback)
  
  return evt.attach(...attachArgs);
```